runtime/yaml.cpp

```diff
@@ -0,0 +1,191 @@
+#include <yaml-cpp/yaml.h>
+
+#include "runtime/optional.h"
+#include "runtime/streams.h"
+#include "yaml.h"
+
+/*
+ * convert YAML::Node to mixed after parsing a YAML document into YAML::Node
+ */
+static void yaml_node_to_mixed(const YAML::Node &node, mixed &data, const string &source) noexcept {
+  data.clear(); // sets data to NULL
+  if (node.IsScalar()) {
+    const string string_data(node.as<std::string>().c_str());
+    // check whether the primitive is put in quotes in the source YAML
+    const bool string_data_has_quotes = (source[node.Mark().pos] == '"' && source[node.Mark().pos + string_data.size() + 1] == '"');
+    // if so, it is a string
+    if (string_data_has_quotes) {
+      data = string_data;
+    } else if (string_data == string("true")) {
+      data = true; // "true" without quotes is boolean(1)
+    } else if (string_data == string("false")) {
+      data = false; // "false" without quotes is boolean(0)
+    } else if (string_data.is_int()) {
+      data = string_data.to_int();
+    } else {
+      double float_data = 0.0;
+      if (string_data.try_to_float(&float_data)) {
+        data = float_data;
+      } else {
+        data = string_data;
+      }
+    }
+  } else if (node.size() == 0 && node.IsDefined() && !node.IsNull()) {
+    // if node is defined, is not null or scalar and has size 0, then it is an empty array
+    array<mixed> empty_array;
+    data = empty_array;
+  } else if (node.IsSequence()) {
+    for (auto it = node.begin(); it != node.end(); ++it) {
+      mixed data_piece;
+      yaml_node_to_mixed(*it, data_piece, source);
+      data.push_back(data_piece);
+    }
+  } else if (node.IsMap()) {
+    for (const auto &it : node) {
+      mixed data_piece;
+      yaml_node_to_mixed(it.second, data_piece, source);
+      data.set_value(string(it.first.as<std::string>().c_str()), data_piece);
+    }
+  }
+  // else node is Null or Undefined, so data is Null
+}
+
+/*
+ * print tabs in quantity of nesting_level (used to print nested YAML entries)
+ */
+static string yaml_print_tabs(const uint8_t nesting_level) noexcept {
+  return string(2 * nesting_level, ' ');
+}
+
+/*
+ * print the key of a YAML map entry
+ */
+static string yaml_print_key(const mixed &data_key) noexcept {
+  if (data_key.is_string()) {
+    return data_key.as_string();
+  }
+  return string(data_key.as_int()); // key can not be an array; bool and float keys are cast to int
+}
+
+/*
+ * escape special characters in a string entry
+ */
+static string yaml_escape(const string &data) noexcept {
+  string escaped_data;
+  for (size_t i = 0; i < data.size(); ++i) {
+    if (data[i] == 10) { // line feed - code 10
```

Comment:
1) extract data[i] to a separate (char) variable: you use in dozen of times
2) why do you compare char with a numeric code, why not `data[i] == '\n' and get rid of comments everywher?